### PR TITLE
[WIP] Immutable network wrapper

### PR DESCRIPTION
**Pull request by geofjamg**
I just started the implementation to validate the design. The goal of an immutable wrapper for the network is to give access to the network via script (for instance contingency config) but to forbid modification.
So question are:
   - is the design ok
   - how to test it, it will add a lot of line of code and to keep a good code coverage we need to traverse all of the wrappers methods. How to share test code of iidm impl with immutable wrapper unit test?

### CODE REVIEW

**Comment by geofjamg** [comment]
Immutable instance mapping cache must not be static because it will cause severe memory leaks. Network and all its components with an immutable wrapper in the actual implementation will never be garbage collected. Caching has to be done with a cache local to each network immutable wrapper.

**Comment by yichen88** [comment]
> Immutable instance mapping cache must not be static because it will cause severe memory leaks. Network and all its components with an immutable wrapper in the actual implementation will never be garbage collected. Caching has to be done with a cache local to each network immutable wrapper.

Wich a cache local, it can not be get the same instance by n.getVL.getSub vs n.getSub.
For memory issue, we could add ImmutableNetwork.release(network) and clean all the static cache maps.

**Comment by geofjamg** [comment]
> > Immutable instance mapping cache must not be static because it will cause severe memory leaks. Network and all its components with an immutable wrapper in the actual implementation will never be garbage collected. Caching has to be done with a cache local to each network immutable wrapper.
> 
> Wich a cache local, it can not be get the same instance by n.getVL.getSub vs n.getSub.
> For memory issue, we could add ImmutableNetwork.release(network) and clean all the static cache maps.

Why not? if you have a common ImmutableCache object share by all wrappers of one ImmutableNetwork instance. You just have to pass ImmutableCache to the constructor of all wrappers.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/ImmutableCacheIndex.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * @author Yichen TANG <yichen.tang at rte-france.com>
+ */
+public class ImmutableCacheIndex {
+
+    private final Map<Identifiable, AbstractImmutableIdentifiable> identifiableCacheMap = new HashMap<>();
```
Comment (geofjamg) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/ImmutableCacheIndex.java]: WeakHashMap needed to avoid memory leak?

**Comment by mathbagu** [comment]
How do you manage extension? It could be understandable to have a read-only network but read-write extensions. What do you think?

**Comment by geofjamg** [comment]
We could have something similar to xml reader/writer extension?
```java
interface ImmutableWrapperExtension {

    <T> T wrap(T extension); 
}
```

**Comment by yichen88** [comment]
> How do you manage extension? It could be understandable to have a read-only network but read-write extensions. What do you think?

Currently it has already supported something like that:
```
        Network network = MultipleExtensionsTestNetworkFactory.create();
        ImmutableNetwork immutableNetwork = ImmutableNetwork.of(network);
        Load load = immutableNetwork.getLoad("LOAD");
        LoadFooExt extension = load.getExtension(LoadFooExt.class);
        extension.setUsername("foo");
        assertEquals("foo", extension.getUsername());
        try {
            load.addExtension(LoadBarExt.class, new LoadBarExt(load));
            fail();
        } catch (Exception e) {
            // ignore
        }
```
If that is what you want, i will add this test in PR.

Or immutable instance could still be added a new extension instead of throwing exception as current implementation?

**Comment by ChamseddineBhd** [comment]
@yichen88 Can you refactor your code to decrease code duplication  ? 
now the duplicated code reached 15.2% on the new Code ( > 3%).
Then can you resolve conflicts and review the 19 code smells ?

**Comment by geofjamg** [comment]
I think we can close this PR as it won't be merged. It adds a lot of code without any clear needs for this feature. Furthermore, we should have implement it using generated proxy (like with bytebuddy) to avoid thousand of hard to maintains line of code.

**Pull request closed without merging**